Multi era json and lint fixes + CI

rooooooooob opened the pull request:
* Add package.json for multi-era wasm building

* Fix up shelley move inst rewards + certs (they are slightly more
restrictive than post-shelley ones that were being used before)

* Change chain/multi-era json-gen crates to be usable as a library too so
we can export all cml-chain JSON schemas from cml-multi-era

* Warnings / clippy lint fixes

* Adds CI for building/testing new crate structure

* impl_wasm_map is now aware of whether the keys/values are WASM-exposable
and copy or not. There is now a away to specify the backing map e.g.
BTreeMap for preserve-encodings=false i.e. Byron, CIP25

* Use of this macro now replaces hard-generated WASM Map (and list) types
to reduce boilerplate + allow a single common place to fix linting
errors (one of the main reasons for this change to be done in this PR -
but it was planned at some point anyway). The rest of clippy linting errors were also fixed and a check in the CI
for fmt + linting.

Fixes #234

Fixes #245

rooooooooob committed:
Multi Era JSON/wasm build fixes

Add package.json for multi-era wasm building

Fix up shelley move inst rewards + certs (they are slightly more
restrictive than post-shelley ones that were being used before)

Change chain/multi-era json-gen crates to be usable as a library too so
we can export all cml-chain JSON schemas from cml-multi-era

rooooooooob committed:
Misc warnings / clippy fixes

rooooooooob committed:
cargo fmt

rooooooooob committed:
Git workflow for new crate structure

* Runs all rust tests

* Runs WASM build process for WASM-terminal crates (chain, multi-era,
  cip25, cip36)

rooooooooob commented:
~TODO: There are a few clippy warnings to fix. Will fix those and add a cargo clippy lint / fmt CI step when I do.~

**Edit: Done**

gostkin commented on .github/workflows/pr-checks.yml:
```diff
@@ -42,3 +42,41 @@ jobs:
       - name: rust:build-browser
         run: |
           npm run rust:build-browser
+      - name: all:test
```
probably it makes sense to have multiple checks, e.g. why to test old cml when only new crates are changed? you can specify the path for particular checks

rooooooooob committed:
impl_wasm_map/list macro rewrite + clippy fixes + CI

impl_wasm_map is now aware of whether the keys/values are WASM-exposable
and copy or not. There is now a away to specify the backing map e.g.
BTreeMap for preserve-encodings=false i.e. Byron, CIP25

Use of this macro now replaces hard-generated WASM Map (and list) types
to reduce boilerplate + allow a single common place to fix linting
errors (one of the main reasons for this change to be done in this PR -
but it was planned at some point anyway)

Fixes #234

The rest of clippy linting errors were also fixed and a check in the CI
for fmt + linting.
